### ISSUE TITLE
IO thread loops infinitely on out event and blocks application with 100% CPU usage [was renamed]

### ISSUE THREAD

**Issue by smattheis** [zeromq/jeromq#520]
**UPDATE: All assumptions in this comment turned out to be wrong. The true problem is described below in comment https://github.com/zeromq/jeromq/issues/520#issuecomment-364570301**

I'm using the latest version 0.4.3 of JeroMQ and my application stops after some time because messages are not sent over a PUSH socket while one of two JeroMQ I/O threads is at 100% CPU usage. The thread remains inside epollWait, see http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/sun/nio/ch/EPollArrayWrapper.java#269

I've seen issue https://github.com/zeromq/jeromq/issues/506 and think it's very likely related. I don't know about the defined behavior of epollWait and how it's supposed to be used in JeroMQ, but I see in my application that the timeout argument is always zero whenever it hangs. As I thought it should be some positive timeout value (*) I replaced `long timeout = executeTimers();` with `long timeout = Math.max(10, executeTimers());` but it doesn't solve the problem and the io thread still hangs forever inside epollWait. Since the thread hangs inside epollWait the selector recreate logic is not executed. Do you have any idea what's wrong?

(*) I saw that JeroMQ implements a similar logic to workaround the JDK bug as netty; however, in netty epollWait is called with a timeout of 10.

In my application, I execute a test that uses four PULL sockets and various PUSH sockets. The test is started multiple times (because I do a bunch of simulations) in the same Java process but each time with a new context. After each test, sockets are closed and the context is closed gracefully.

**Comment by fredoboulo** [comment]
Hi, could you provide some tests for us to get a better understanding of your situation? Some information about your OS, java version , ... would be useful as well. Does it occur when loosing network, hibernating computer, ...?

**Comment by smattheis** [comment]
In my test the error occurs with 90% probability in a 5 min execution time such that I can actually reproduce the problem quite reliably. However, I can't share that code. I tried to reproduce the problem in an isolated test but without success. Nevertheless, I will keep on trying to reproduce it.

In my case, I think the problem is not related (directly) to hibernating or network issues. (I have read that in issue https://github.com/zeromq/jeromq/issues/506 it's reported to be related to hibernating.) Instead, I have a bunch of worker threads that perform compute-intense calculations with lots of data in main memory and exchange messages to synchronize their work. I observed (but not 100% sure) that it usually occurs if the application exceeds memory and jvm starts swapping. This could be related to hibernating under Linux because  AFAIK Linux just moves memory to swap if it hibernates.

For the records:
``` bash
$ uname -a
Linux ubuntu 4.4.0-112-generic #135-Ubuntu SMP Fri Jan 19 11:48:36 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
$ java -version
Picked up _JAVA_OPTIONS: -Xmx10g -Xss8m
java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)
```
Note: Above you also see the heap size that I use on am machine with 12gb main memory and a quad-core with SMT, so 8 virtual cores.

**Comment by fredoboulo** [comment]
The way to rebuild selector due to the epoll JDK bug may be in cause, could you add some traces to see if counter is correctly incremented and if the selector is being rebuilt, or is the code just totally blocked in epollWait ?

This 100 % CPU on epollWait created some litterature, with no clear workaround apart from [netty solution](https://github.com/netty/netty/issues/327). I also found some discussions about possible collision with garbage collector, [especially here](https://stackoverflow.com/questions/28662737/weblogic-socket-muxer-uses-100-cpu). I do not like this idea much, and it was flagged as not changing the situation but... Would you mind make it a try, and start your app with the extra settings?

> -XX:+UseConcMarkSweepGC 
> -XX:+UseParNewGC 
> -XX:ParallelCMSThreads=2 
> -XX:+CMSParallelRemarkEnabled 
> -XX:+CMSIncrementalMode 
> -XX:+CMSIncrementalPacing 
> -XX:CMSFullGCsBeforeCompaction=1 
> -XX:+CMSClassUnloadingEnabled 
> -XX:CMSInitiatingOccupancyFraction=80

Sorry if I cannot make it myself, I lack occurrences of this class of problem on my machines.

**Comment by smattheis** [comment]
Okay, I did some traces and found that my previous assumption was completely wrong. I think it's a bug in the BaseEncoder: It seems that the outpos buffer is corrupted such that its size is zero while outsize remains greater than zero. As a consequence, the io thread loops over the out event over and over again because outsize is greater than zero but without sending anything because the buffer is empty. I tried to trace the problem back to its root cause and I think it's because out event flips the buffer after filling if and only if its size is less or equals OUT_BATCH_SIZE, see https://github.com/zeromq/jeromq/blob/43463e02af3375e83d10c0562aaa69b9d21f9472/src/main/java/zmq/io/StreamEngine.java#L576 AFAIK the condition is to avoid flipping if the EncoderBase#encode() performed a zero-copy encoding because the writeBuf of the encoder is already flipped. - Is that correct? - However, a problem occurs if toWrite equals bufferSize, which is set to OUT_BATCH_SIZE, see https://github.com/zeromq/jeromq/blob/bb0428c4d16d6c6045de303d9a43ad39239704d3/src/main/java/zmq/io/coder/EncoderBase.java#L95 because encode applies a zero-copy encoding but its later flipped in the out event. That means the condition for flipping in the outEvent is ambiguous for a buffer size of OUT_BATCH_SIZE. I tried a fix where I added a condition in the EncoderBase `toWrite > OUT_BATCH:SIZE` and did not change the condition `toWrite >= bufferSize` to avoid side effects if other parts of JeroMQ use a different bufferSize. That fixes solves the problem for me. I even traced the error condition to verify that it would occur (in a very long execution) and the fix prevented the buffer from being corrupted.
The bugfix is included in PR https://github.com/zeromq/jeromq/pull/522

**Comment by fredoboulo** [comment]
Hello,

thanks for tackling that down! Codebase of StreamEngine is different from [C++ equivalent](https://github.com/zeromq/zeromq4-1/blob/a3f98304da9024f2d0ac4e838a646c231ddff5ee/src/stream_engine.cpp#L377), maybe there is something to dig in that part of the code.

I'm not sure about the modification in EncoderBase, the [comparison is >= in zeromq4-1](https://github.com/zeromq/zeromq4-1/blob/a3f98304da9024f2d0ac4e838a646c231ddff5ee/src/encoder.hpp#L113).

I really wish we got a unit test for that part of the code. I see some usage of it (some tests like ByteBuffersTest do not flip the buffer), but it may not be enough.

**Comment by smattheis** [comment]
Okay, the problem is just that because of the flipping of java.nio.ByteBuffer's JeroMQ requires a different behavior than ZeroMQ with plain C++ byte arrays. The logic and therefore also the code cannot be the same.
What do you intend to do with the bug and the PR? All I know is that it's actually a serious bug that completely blocks an application even though it occurs with very low probability. As shown, the solution is trivial for the described case. It's rather a question how the maintainers want to organize such things and, respectively, such pull requests:

Is it a high requirement for the maintainers to keep Java code as close to the C++ ZeroMQ code as possible for simplified maintenance which, however, may require some well defined interfaces between JeroMQ's C++-equivalent code and JRE-specific code and certainly comes with some downsides with regard to performance? Or, should it be optimized for performance on JRE which requires much more divergence from C++ code? I would totally understand and follow either approach but if I continue working/contributing with/to JeroMQ I would need to know that. :) I can try to revise the PR and follow your guidelines, just let me know.

**Comment by fredoboulo** [comment]
Sorry, no time for me to be on this topic this week.

I will be fast now and will provide more context later, but as long as your PR satisfies the [terms of C4](https://rfc.zeromq.org/spec:42/C4), it will be merged.

The following is not the project's thinkings but my very own:
when the lib was bumped to 4.1.7, I decided to stick as much as possible to the logic of libzmq so it would be better to maintain. Not everything can be translated back to Java (ByteBuffer or equivalents are not used in C++ version), but the more the code sticks to it, the better I feel. To answer your question, I personally would go for both in the given order:

1. make code work, with good performances
2. be as close as possible to C++ version

In your PR, this comparison in EncoderBase makes a deviation compared to the C++ version, which I find hard to explain, while at the same time a code that is present only in Java to decide flipping the buffer or not seems to be the source of the bug you reported. If I was you, I would invest a bit of time to try to refine that Java-specific code. But I am not you :)

If you can wait a week, I may be more available then.

**Comment by smattheis** [comment]
Thanks for the explanation, that helps a lot. I'm not in a hurry to get that merged and sorry, if I rushed you.

Since the bug is due to an ambiguity of ByteBuffer flipping if buffer size equals out batch size, which in turn is because buffer size is (mis)used for this logic instead of a dedicated parameter, there are different ways to resolve the ambiguity. That means, it could also be resolved in the StreamEngine instead of the EncoderBase, or even with an additional parameter.

As said, I'm available to revise the PR and also for discussion once you have time to think about it and decide which strategy would be best for maintenance. For now, I keep this issue and the PR open.

**Comment by fredoboulo** [comment]
Hi @smattheis could you have a look  at the commit I just made and tell me if it solves the problem?

**Comment by fredoboulo** [comment]
Hello @smattheis , any news regarding that issue?

**Comment by fredoboulo** [comment]
Hi, thanks to you for spotting this bug. I will submit the PR without waiting so much, to gather feedback from others. Please tell if this solves your problem, and do not hesitate to re-open the PR if it continues, as merging it will automatically close this issue.

**Comment by smattheis** [comment]
To be honest, it looks like it doesn't solve the bug because it doesn't resolve the ambiguity. If you give me a bit more time, I will check it out and will also verify if the tests actually cover the problematic scenario. Maybe I'm wrong and it solves the bug. (I haven't had the time to work on it last week.)

**Comment by fredoboulo** [comment]
Fine, I put my PR on stall for now, to give you time to check it.

Just for you to be aware, I added a unit test for the situation, you can review it from that side as well.

The big thing is that your own PR has some failing tests which is a no-go-no-merge situation. Plus the fact that changing condition from > to >= makes me uncomfortable, but it's more of a personal thing.

**Comment by smattheis** [comment]
Okay, I checked it and let my application run through the problem scenario. It resolves the issue. Sorry for my doubts, I was not aware that your commit resolves the ambiguity through the reference of outpos that usually points to encoder's buffer but in the "shortcut scenario" to the encoder's writeBuf such that the ByteBuffer used for sending is not flipped in the problem scenario because encoded only flips the encoder's buffer. I also checked the test and found that it provokes the problem scenario various times. In conclusion, it's all perfect. Thanks!
